### ISSUE TITLE
IE8 Support - Broken by transpilation

### ISSUE THREAD

**Issue by rwjblue** [emberjs/ember.js#10730]
Our compatibility with IE8 has been broken by the recent few pull requests adding ES6 features.

---

Example (destructuring):

Input:

```javascript
function foo(blah) {
 var [a,b] = blah; 
}
```

Output:

```javascript
"use strict";

var _slicedToArray = function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { var _arr = []; for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) { _arr.push(_step.value); if (i && _arr.length === i) break; } return _arr; } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } };

function foo(blah) {
  var _blah = _slicedToArray(blah, 2);

  var a = _blah[0];
  var b = _blah[1];
}
```

---

Example (spread):

Input:

```javascript
function doIt(foo) {
  blah(...foo)
}
```

Output:

```javascript
"use strict";

var _toConsumableArray = function (arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } };

function doIt(foo) {
  blah.apply(undefined, _toConsumableArray(foo));
}
```
---

The above examples show usage of destructuring and spread, but Babel currently assumes `Array.isArray` is present (which is not true of IE8).

**Comment by rwjblue** [comment]
We need to revert usage of both destructuring and spread to fix the build on IE8.

**Comment by rwjblue** [comment]
@danmcclain - You mentioned you might have time, could you take a look at this?

**Comment by danmcclain** [comment]
@rwjblue I can dig into this

**Comment by et** [comment]
Is there a tracked ticket for fixing this in Babel? I had a similar issue when using spread for a project.

**Comment by rwjblue** [comment]
@et - I don't know that they consider it a bug, but it was initially reported in https://github.com/babel/babel/issues/1055 (though that issue has somewhat morphed from its goal).

**Comment by wycats** [comment]
`Array.isArray` is pretty easy to polyfill. Can we just do that?

**Comment by rwjblue** [comment]
@wycats - Yes, we could polyfill (and have considered it) but IMHO we (as a framework) should not be mutating platform classes/objects.

**Comment by sebmck** [comment]
Next major of Babel will use `arr && arr.constructor === Array` instead of `Array.isArray`.

@rwjblue Nothing related to this was reported in  babel/babel#1055.

**Comment by stefanpenner** [comment]
> arr && arr.constructor === Array

I believe isn't cross realm safe

**Comment by sebmck** [comment]
@stefanpenner Good point.

**Comment by sebmck** [comment]
Ok so in the next Babel major you can enable [loose mode](http://babeljs.io/docs/usage/loose/) for the `es6.destructuring` and `es6.spread` transformers and it'll assume all values are arrays and ignore iterables so the helpers wont be used and direct references instead. ie.

```javascript
function foo(blah) {
 var [a,b] = blah; 
}
```

becomes

```javascript
function foo(blah) {
  var a = blah[0];
  var b = blah[1];
}
```

Should be enough to cover your use-case.

**Comment by stefanpenner** [comment]
@sebmck has also informed me that this available in babel#experimental as of minutes ago.

**Comment by rwjblue** [comment]
@sebmck - Thank you! It was my understanding (from chatting with Stef before he opened the issue) was that the point of https://github.com/babel/babel/issues/1055 was essentially the issue I reported here.  I am sorry for my confusion.

**Comment by MajorBreakfast** [comment]
I think you shouldn't worry about polyfilling stuff like that. If browser vendors think that they can add this API without breaking the web then it's certainly safe if ember does it.

**Comment by rwjblue** [comment]
Fixed by #10816.
